File: src/android/firebase_performance/src/main/java/com/ee/FirebasePerformanceBridge.kt
package com.ee

import android.app.Activity
import android.app.Application
import androidx.annotation.AnyThread
import com.ee.internal.FirebasePerformanceTrace
import com.google.firebase.ktx.Firebase
import com.google.firebase.perf.FirebasePerformance
import com.google.firebase.perf.ktx.performance
import java.util.concurrent.ConcurrentHashMap

class FirebasePerformanceBridge(
    private val _bridge: IMessageBridge,
    private val _logger: ILogger,
    private val _application: Application,
    private var _activity: Activity?) : IPlugin {
    companion object {
        private val kTag = FirebasePerformanceBridge::class.java.name
        private const val kPrefix = "FirebasePerformanceBridge"
        private const val kInitialize = "${kPrefix}Initialize"
        private const val kIsDataCollectionEnabled = "${kPrefix}IsDataCollectionEnabled"
        private const val kSetDataCollectionEnabled = "${kPrefix}SetDataCollectionEnabled"
        private const val kNewTrace = "${kPrefix}NewTrace"
    }

    private var _plugin: FirebasePerformance? = null
    private val _traces: MutableMap<String, FirebasePerformanceTrace> = ConcurrentHashMap()

    init {
        _logger.info("$kTag: constructor begin: application = $_application activity = $_activity")
        registerHandlers()
        _logger.info("$kTag: constructor end.")
    }

    override fun onCreate(activity: Activity) {
        _activity = activity
    }

    override fun onStart() {}
    override fun onStop() {}
    override fun onResume() {}
    override fun onPause() {}

    override fun onDestroy() {
        _activity = null
    }

    override fun destroy() {
        deregisterHandlers()
        for (trace in _traces.values) {
            trace.destroy()
        }
        _traces.clear()
    }

    @AnyThread
    private fun registerHandlers() {
        _bridge.registerAsyncHandler(kInitialize) {
            Utils.toString(initialize())
        }
        _bridge.registerHandler(kIsDataCollectionEnabled) {
            Utils.toString(isDataCollectionEnabled)
        }
        _bridge.registerHandler(kSetDataCollectionEnabled) { message ->
            isDataCollectionEnabled = Utils.toBoolean(message)
            ""
        }
        _bridge.registerHandler(kNewTrace) { message ->
            Utils.toString(newTrace(message))
        }
    }

    @AnyThread
    private fun deregisterHandlers() {
        _bridge.deregisterHandler(kInitialize)
        _bridge.deregisterHandler(kSetDataCollectionEnabled)
        _bridge.deregisterHandler(kIsDataCollectionEnabled)
        _bridge.deregisterHandler(kNewTrace)
    }

    @AnyThread
    private suspend fun initialize(): Boolean {
        if (FirebaseInitializer.instance.initialize(false)) {
            _plugin = Firebase.performance
            return true
        }
        return false
    }

    private var isDataCollectionEnabled: Boolean
        @AnyThread get() {
            val plugin = _plugin ?: throw IllegalStateException("Please call initialize() first")
            return plugin.isPerformanceCollectionEnabled
        }
        @AnyThread set(value) {
            val plugin = _plugin ?: throw IllegalStateException("Please call initialize() first")
            plugin.isPerformanceCollectionEnabled = value
        }

    @AnyThread
    fun newTrace(traceName: String): Boolean {
        val plugin = _plugin ?: throw IllegalStateException("Please call initialize() first")
        if (_traces.containsKey(traceName)) {
            return false
        }
        val trace = plugin.newTrace(traceName)
        _traces[traceName] = FirebasePerformanceTrace(_bridge, trace, traceName)
        return true
    }
}

File: src/ios/ee/firebase_performance/FirebasePerformanceBridge.swift
//
//  FirebasePerformanceBridge.swift
//  ee-x-bb8dc50f
//
//  Created by eps on 2/3/21.
//

private let kPrefix = "FirebasePerformanceBridge"
private let kInitialize = "\(kPrefix)Initialize"
private let kIsDataCollectionEnabled = "\(kPrefix)IsDataCollectionEnabled"
private let kSetDataCollectionEnabled = "\(kPrefix)SetDataCollectionEnabled"
private let kNewTrace = "\(kPrefix)NewTrace"

@objc(EEFirebasePerformanceBridge)
class FirebasePerformanceBridge: NSObject, IPlugin {
    private let _bridge: IMessageBridge
    private var _plugin: Performance?
    private var _traces: [String: FirebasePerformanceTrace]
    
    public required init(_ bridge: IMessageBridge, _ logger: ILogger) {
        _bridge = bridge
        _plugin = nil
        _traces = [:]
        super.init()
        registerHandlers()
    }
    
    public func destroy() {
        deregisterHandlers()
    }
    
    func registerHandlers() {
        _bridge.registerAsyncHandler(kInitialize) { _, resolver in
            resolver(Utils.toString(self.initialize()))
        }
        _bridge.registerHandler(kIsDataCollectionEnabled) { _ in
            Utils.toString(self.isDataCollectionEnabled)
        }
        _bridge.registerHandler(kSetDataCollectionEnabled) { message in
            self.isDataCollectionEnabled = Utils.toBool(message)
            return ""
        }
        _bridge.registerHandler(kNewTrace) { message in
            Utils.toString(self.newTrace(message))
        }
    }
    
    func deregisterHandlers() {
        _bridge.deregisterHandler(kInitialize)
        _bridge.deregisterHandler(kIsDataCollectionEnabled)
        _bridge.deregisterHandler(kSetDataCollectionEnabled)
        _bridge.deregisterHandler(kNewTrace)
    }
    
    func initialize() -> Bool {
        if FirebaseInitializer.instance.initialize() {
            _plugin = Performance.sharedInstance()
            return true
        }
        return false
    }
    
    var isDataCollectionEnabled: Bool {
        get {
            guard let plugin = _plugin else {
                assert(false, "Please call initialize() first")
                return false
            }
            return plugin.isDataCollectionEnabled
        }
        set(value) {
            guard let plugin = _plugin else {
                assert(false, "Please call initialize() first")
                return
            }
            return plugin.isDataCollectionEnabled = value
        }
    }
    
    func newTrace(_ traceName: String) -> Bool {
        guard let plugin = _plugin else {
            assert(false, "Please call initialize() first")
            return false
        }
        if _traces.contains(where: { key, _ in key == traceName }) {
            return false
        }
        guard let trace = plugin.trace(name: traceName) else {
            return false
        }
        _traces[traceName] = FirebasePerformanceTrace(_bridge, traceName, trace)
        return true
    }
}
